[#7, #114] 카테고리 삭제 기능 구현, 카테고리  페이지 구현

gomster96 opened the pull request:
### Description
- 카테고리 삭제 기능을 구현하였다.
- 카테고리 생성 기능을 하는 관리자 페이지를 구현하였다.
- 카테고리 삭제 기능을 하는 관리자 페이지를 구현하였다.

### Trouble Shooting
- Thymeleaf에서 form 으로 deleteMethod를 사용하고 싶은데 다른 블로그들에서 단순히 hidden input만 만들어 준것으로 해결이 되지 않았습니다. application.yml 에 `spring.mvc.hiddenmethod.filter.enabled=true`조건을 추가하여 @DeleteMapping 을 hidden input을 통해 사용할 수 있도록 해결하였습니다. [참고링크](https://stackoverflow.com/questions/24256051/delete-or-put-methods-in-thymeleaf)
### ETC
- 부모 카테고리를 삭제 했을 때 부모카테고리의 자식 카테고리까지 모두 삭제 처리를 해주었는데, 이들의 관계는 일부로 끊지 않았습니다. 어떻게 하는게 더 나을지 고민입니다.
-  ~~자식 카테고리를 삭제 했을 경우에는 부모카테고리와의 관계를 끊어주었습니다~~
- 카테고리를 삭제할 때 parentCategory가 삭제되면 해당 category를 지닌 shop들도 모두 삭제하였습니다.

gomster96 committed:
feat[#7] : Category 생성 service 구현

gomster96 committed:
refactor[#7] : Category 생성시 사용하던 requestDto 이름 변경 

category 단일 조회할 때 사용하는 dto를 CategoryRequestDto로 이름지엇기때문에
categoryRequestDto -> CategoryCreateRequestDto 변경

gomster96 committed:
feat[#7] : Category Controller 단 구현

gomster96 committed:
refactor[#7] : dto 이름 변경

- 기존 categoryReqeustDto 는 create를 위한 request였으므로 CategoryCreateRequestDto로 이름 변경

gomster96 committed:
feat[#7] : category delete service 기능 구현 

- delete 를 연관관계 편의 메소드를 만들어, 연관관계를 제거해주며, 자식이 제거 될 경우, 부모와의 연관관계를 끊어주었다.
- 부모가 제거 될 경우, 모든 자식이 제거되도록 구현하였다. 
- @Where 를 사용하여 soft delete 된 상태의 데이터는 조회되지 않도록 하였다.

gomster96 committed:
test[#7] : category delete integration test

gomster96 committed:
feat[#114] : 카테고리 생성, 삭제 Admin Controller 기능 구현

gomster96 committed:
feat[#114] : 카테고리 생성, 삭제 관리자 페이지 기능 구현

gomster96 committed:
feat[#114] : @DeleteMapping 작동 활성화

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -70,4 +73,13 @@ public List<RootCategoryResponseDto> findRootCategoriesAndChildren() {
+
+    @Transactional
+    public void delete(CategoryRequestDto categoryRequestDto){
+        Category category = categoryRepository.findById(categoryRequestDto.getId())
+                                              .orElseThrow(CategoryIdException::new);
```
위에 findById라는 메서드가 있는데 해당 메서드를 호출하는 건 어떨까요? 로직도 완전히 동일한 것 같습니다

KJunseo replied:
추가적으로 예외이름을 `NoSuchCategoryException`이나 `CategoryNotFoundException`같이 많이 쓰이는 형태로 만드는 건 어떤가요?

gomster96 replied:
아 맞습니다.  다만 서비스의 findById를 현제 테스트에서만 사용하기 때문에 삭제될 수도 있다고 생각해서 그렇게 하였지만, 지금 생각해보니 말씀하신대로 이미 만들어진 것을 사용하는것이 나을거 같네요. 감사합니다.

많이 쓰이는 Exception의 형태로 바꾸는게 훨씬 사용범위나 코드 가독성이나 좋겠네요. 감사합니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -70,4 +73,13 @@ public List<RootCategoryResponseDto> findRootCategoriesAndChildren() {
+    public void delete(CategoryRequestDto categoryRequestDto){
+        Category category = categoryRepository.findById(categoryRequestDto.getId())
+                                              .orElseThrow(CategoryIdException::new);
+        if(category.isDeleted()) throw new CategoryAlreadyDeletedException();
+        category.getChildren().forEach(BaseEntity::deleteContent);
+        category.delete();
+    }
```
if문이 하나여도 가독성과 일관성을 위해 {}를 붙여주는 게 좋을 것 같습니다.

KJunseo replied:
예외를 던지는 것을 delete 메서드 안에서 처리해도 되지 않을까요? 현재는 객체에서 값을 꺼내서 비교하는 것 같습니다. 개인적으로 객체에 메세지를 보내는 게 좀 더 좋아보여요

KJunseo replied:
카테고리에 포함된 서브카테고리들을 삭제하는 과정도 카테고리에서 데이터를 꺼내서 처리하기 보다 메세지를 보내 내부에서 처리하는 게 좋아보입니다.

gomster96 replied:
네 감사합니다. 카테고리에서 객체 내부에서 메시지를 받아 처리하도록 구현하겠습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Category.java:
```diff
@@ -66,6 +69,19 @@ public boolean isSameName(String name) {
+        category.parent = null;
+    }
+
+    public void delete() {
+        if (parent != null) {
+            parent.deleteChildCategory(this);
+            parent = null;
+        }
+        deleteContent();
+    }
```
현재 서브 카테고리를 삭제할 때 `deleteChildCategory`를 호출하는데 `deleteChildCategory`안에서도 부모 카테고리를 null로 만들고, 80번째 줄에서도 null로 만드는데 2번 해주시는 이유가 있나요?

KJunseo replied:
soft delete를 사용하는 이유중 하나가 복구가 가능하다는 점인 것 같은데, 실제로 parent를 null로 만들면 복구가 불가능해지지 않나요?

모든 엔티티에 soft delete를 적용할 필요는 없다고 생각하지만, soft delete를 적용할거라면 deleted 필드값만 변경해주는 게 맞다고 생각해요!

KJunseo replied:
추가적으로 카테고리를 삭제처리 했을 때 해당 카테고리에 포함되는 가게가 존재할 경우 어떻게 처리할지도 정해서 처리를 해줘야 할 것 같습니다.

삭제된 카테고리로 그대로 보여줄 수는 없을 것 같아요. 지금 당장 생각나는 로직은 

삭제하려는 카테고리에 포함된 가게가 있는 경우 -> 임시로 새로운 카테고리를 만들어서 배정 -> 삭제하려는 카테고리에 포함된 가게 수가 0이 되었으면 삭제 

이건데 회의때 어떻게 처리할지 정해서 처리해주는 게 좋아보이네요

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -70,4 +73,13 @@ public List<RootCategoryResponseDto> findRootCategoriesAndChildren() {
     }
+
+    @Transactional
+    public void delete(CategoryRequestDto categoryRequestDto){
```
```suggestion
    public void delete(CategoryRequestDto categoryRequestDto) {
```
포매팅을 적용해주세요!

KJunseo commented on backend/src/main/java/com/handong/rebon/category/presentation/AdminCategoryController.java:
```diff
@@ -0,0 +1,67 @@
+    @GetMapping("/categories/new")
+    public String categoryCreateForm(Model model){
+
+        List<RootCategoryResponseDto> categories = categoryService.findRootCategoriesAndChildren();
```
apiController와 같이 `RootCategoryResponse`로 맞춰주는 건 어떤가요?

KJunseo commented on backend/src/main/java/com/handong/rebon/category/presentation/AdminCategoryController.java:
```diff
@@ -0,0 +1,67 @@
+
+    private final CategoryService categoryService;
+
+    @GetMapping("/categories/new")
+    public String categoryCreateForm(Model model){
```
현재 최상위 카테고리를 만드는 부분은 보이지 않는데 추가하실 계획이신가요?

gomster96 replied:
최상위 카테고리의 경우 부모값을 null로 주면 생성되도록 만들었습니다..!

KJunseo commented on backend/src/main/java/com/handong/rebon/category/presentation/AdminCategoryController.java:
```diff
@@ -0,0 +1,67 @@
+    public String createCategory(CategoryCreateRequest categoryRequest){
+
+        CategoryCreateRequestDto categoryCreateRequestDto = CategoryAssembler.categoryCreateRequestDto(categoryRequest);
+        if(categoryCreateRequestDto.getParentId() == null){
+            categoryService.create(categoryCreateRequestDto.getName());
+        } else {
+            categoryService.create(categoryCreateRequestDto);
+        }
```
저희 컨벤션에 따라 else 예약어는 쓰지 않는게 좋아보여요

추가적으로 데이터를 꺼내서 null 체크하는 것도 로직이라는 생각이 드는데 controller 단에는 로직이 들어가지 않아야한다고 생각해요

gomster96 replied:
네 알겠습니다 else예약어를 제가 자꾸쓰네요.. 주의하겠습니다. 그리고 null을 꺼내는 것도 로직이기에 서비스 단으로 내리도록 하겠습니다.

KJunseo commented on backend/src/main/resources/application.yml:
```diff
@@ -13,6 +13,10 @@ spring:
       hibernate:
         format_sql: true
         default_batch_fetch_size: 100
+  mvc:
+    hiddenmethod:
+      filter:
+        enabled: true
```
오 이 설정은 처음보네요.

KJunseo reviewed (changes requested):
궁금한 점과 몇가지 코멘트 남겼습니다!
확인 후 재요청 부탁드려요!

- 현재 카테고리 부분은 아니지만 ReviewDeleteIntegrationTest 테스트가 동작하지 않습니다. 아마 BaseEntity를 수정하면서 그렇게 된 것 같은데 pr전에 테스트 잘 돌아가는지 확인부탁드립니다!
- Category에 로직이 추가되었는데 unit 테스트가 작성되지 않은 것 같습니다!
- pr전 코드 컨벤션과 포매팅 한 번만 확인 부탁드립니다!

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/category/presentation/AdminCategoryController.java:
```diff
@@ -0,0 +1,67 @@
+        return "category/createForm";
+    }
+
+    @GetMapping("/categories/remove")
+    public String categoryDeleteForm(Model model){
+
+        List<RootCategoryResponseDto> categories = categoryService.findRootCategoriesAndChildren();
+        model.addAttribute("categories", CategoryAssembler.rootCategoryResponses(categories));
+
+        return "category/deleteForm";
+    }
```
궁금한데, deleteForm이 따로 필요한거요??
카테고리들을 다 불러오는 페이지에서 delete 버튼을 추가해서 삭제하는 거는 어떤가요?

gomster96 replied:
사실 그렇게 해도 되긴 하는데 조금 더 빠른 개발을 위해서 했습니다...!

PyeongGangKim reviewed (changes requested):
고생하셨습니다ㅎㅎ
궁금한 내용 있어서, 리뷰 남겼어요!!
준서님이 리뷰를 잘해주셔서, 그것들 확인하시고 다시 리뷰 요청하시면 될 거 같아요!!
추가적으로, 어드민에 의한 리뷰 삭제, 리뷰 작성자가 자신의 리뷰 삭제 테스트가 안돌아가는 데, 이건 머지 하실때, 해결할 예정이신가요??

gomster96 committed:
refactor[#7] : CategoryService 

- delete 메서드에서 이미 구현해둔 findById와 로직이 겹치는 부분이 있어서 이를 제거했다.
- Exception의 이름을 보다 잘 쓰이는 CategoryNotFoundException으로 변경했다.

gomster96 committed:
refactor[#7] : 카테고리를 삭제 쿼리변경

- 기존에는 fetch join하고 이후 관계를 끊어주는 delete를 하였지만, 논의 결과 관계는 유지하는 soft delete를 하기로 변경되었으므로, batch size가 고정되어있기 때문에 N+1문제는 발생하지 않으므로 이렇게 처리하였다.

gomster96 committed:
refactor[#7] : 카테고리 service delete 일부 수정 , AdminCategorycontroller 로직 없애기

- category delete 에서 delete하는 로직을 category에 메세지를 보내 category가 수정하도록 구현하였다 .
- AdminCategoryController에 비즈니스 로직이 있는 것을 categoryService로 내려주었다.

gomster96 committed:
refactor[#7] : 기존에 만들어있는 get 메서드 사용

gomster96 committed:
test[#7] : 카테고리 삭제 unit 테스트 추가

gomster96 committed:
test[#7] : 카테고리 삭제 통합테스트 수정

- entityManager 의 flush, clear를 사용하여 삭제된 후 다시 조회 하는 것에 있어서 영속성 컨텍스트에 남는것을 방지하였다. 
- CategoryNotFoundException 으로 변경하였다.

gomster96 committed:
refactor[#7] : review test에서 에러 수정

- isDeleted로 되어있어서 extract하지 못하던 에러를 다시 deleted로 수정하여 바꿔주었다.

gomster96 committed:
merge from dev

gomster96 committed:
refactor[#7] : 머지 후 포멧팅 적용

gomster96 committed:
feat[#7] : 루트 카테고리가 삭제되었을 때 관련된 가게 모두 삭제

- 기존의 shop의 addCategories메서드에서 부모에 대한 연관관계메서드가 없었으므로 이를 추가해주었다. 
- 부모카테고리가 null인 즉 root카테고리를 삭제할 때 모든 식당 도 같이 삭제되도록 구현하였다.

gomster96 committed:
test[#7] : 루트 카테고리 삭제시 해당 카테고리의 shop 모두 삭제 테스트

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -34,12 +38,14 @@ public Long create(String categoryName) {
-
-        Category parent = categoryRepository.findById(categoryRequestDto.getParentId())
+    public Long create(CategoryCreateRequestDto categoryCreateRequestDto) {
+        if (categoryCreateRequestDto.getParentId() == null) {
```
== null 대신 Objects.isNull()은 어떤가요?

gomster96 replied:
가독성을 위해 반영하겠습니다.!

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -34,12 +38,14 @@ public Long create(String categoryName) {
+        if (categoryCreateRequestDto.getParentId() == null) {
+            return this.create(categoryCreateRequestDto.getName());
+        }
+        Category parent = categoryRepository.findById(categoryCreateRequestDto.getParentId())
```
기존에 존재하는 findById()를 사용하지 않으신 이유가 있나요?

gomster96 replied:
아... 이거 사용하는걸로 바꾸었었는데 merge하고 다시 이전꺼로 되돌아갔던거같아요..! 흠 다시 확인해서 올리겠습니다.!

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -55,6 +61,23 @@ private void checkCategoryExist(String name) {
+    @Transactional
+    public void delete(CategoryRequestDto categoryRequestDto) {
+        Category category = this.findById(categoryRequestDto.getId());
+        if (category.getParent() == null) {
+            List<ShopCategory> shopCategories = category.getShopCategories();
+            this.deleteAllShopOfCategory(shopCategories);
+        }
```
category 내부에 메세지를 보내 처리하는 건 어떤가요? 현재는 parent도 꺼내고, shopCategories도 꺼내서 로직이 수행되어 객체지향적이지 않은것 같아요

gomster96 replied:
네 저도 그게 맞는것 같습니다..! 메세지를 보내서 해당 객체가 로직을 수행하도록 바꾸겠습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -55,6 +61,23 @@ private void checkCategoryExist(String name) {
+    }
+
+    @Transactional
+    public void deleteAllShopOfCategory(List<ShopCategory> shopCategories) {
+        shopCategories.forEach(shopCategory -> {
+            shopRepository.delete(shopCategory.getShop());
+        });
```
명시적으로 delete 하지 않고, 객체에서 삭제하는 식으로 처리하는건 어떤가요?

gomster96 replied:
네 저도 그게 맞다고 생각합니다! 자꾸 고려하지 못하네요!! 감사합니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Categories.java:
```diff
@@ -23,4 +23,8 @@ public void addChild(Category category) {
     private void checkDuplicateCategory(Category category) {
         this.categories.forEach(child -> child.validateSame(category));
     }
+
+    public void deleteChild(Category category){
+        this.categories.remove(category);
+    }
```
사용되지 않는 메서드인것 같아요

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Category.java:
```diff
@@ -66,6 +70,14 @@ public boolean isSameName(String name) {
+        if (isDeleted()) {
+            throw new CategoryAlreadyDeletedException();
+        }
+        this.getChildren().forEach(BaseEntity::deleteContent);
```
일급컬렉션을 쓰고 있기 때문에 get으로 가져와서 처리하기 보다 메세지를 보내는 게 좋아보여요

gomster96 replied:
맞습니다...!

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Category.java:
```diff
@@ -66,6 +70,14 @@ public boolean isSameName(String name) {
         return this.name.equals(name);
     }
 
+    public void delete() {
```
저번 회의 때 
- 상위 카테고리: 가게 soft delete
- 하위 카테고리: 연관관계 soft delete

로 기억을 하는데 하위 카테고리 삭제시 shopCategory도 soft delete되고 있나요??

gomster96 replied:
네 관련해서 테스트도 같이 구현하겠습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/Shop.java:
```diff
@@ -92,6 +92,7 @@ public void addTags(List<Tag> tags) {
 
     public void addCategories(Category parent, List<Category> subCategories) {
         this.category = parent;
+        parent.getShopCategories().add(new ShopCategory(this, parent));
```
이 부분은 어떤 역할을 하는 건가요? shopCategory는 최상위 카테고리 정보는 가지고 있지 않은 테이블입니다!

PyeongGangKim replied:
저도 이 부분이 왜 추가되어야 하는지 이해가 되지 않습니다!!
혹시 설명 한번 부탁드려도 될까요?

gomster96 replied:
이렇게 했을 때 parent Category의 경우 자신이 가지고 있는 shopcategory들에 대한 정보가 없습니다. 이렇게 되면 rootCategory를 삭제할때 가지고 있는 모든 shop을 삭제해야하는데 가지고 있는 shop의 정보가 없기 때문에 그렇습니다. 만약에 이렇게 하지 않으려면 shop에서 "해당 카테고리를 가지고 있는 shop을 삭제한다" 이런식의 쿼리를 만들어줘야하는데, 이것보다는 이렇게 parent 도 자기가 가지고 있는 shop의 정보를 가지고 있는게 나을것 같아서 이렇게 관계를 맺어주었습니다.

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryDeleteIntegrationTest.java:
```diff
@@ -0,0 +1,165 @@
+    private ShopRepository shopRepository;
+
+    @Test
+    @DisplayName("자식 카테고리를 삭제한다.")
+    public void 자식_카테고리_삭제() {
```
shopCategory 테이블도 soft delete 되었는지 검증하면 좋을 것 같습니다.

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryDeleteIntegrationTest.java:
```diff
@@ -0,0 +1,165 @@
+        categoryService.delete(requestDto);
+        entityManager.flush();
+        entityManager.clear();
+        //        categoryRepository.flush();
```
지우는게 좋을 것 같아요!

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryDeleteIntegrationTest.java:
```diff
@@ -0,0 +1,165 @@
+                .isInstanceOf(CategoryNotFoundException.class);
+    }
+
+    @Test
+    @DisplayName("루트 카테고리를 삭제할 경우, 해당 카테고리에 포함된 모든 Shop을 삭제한다.")
+    public void 루트_카테고리_상점_삭제() {
```
이 메서드가 너무 길어 가독성이 떨어지는 것 같아요. 개행과 메서드 분리를 하면 좋을 것 같습니다.

KJunseo commented on backend/src/test/java/com/handong/rebon/integration/category/CategoryReadIntegrationTest.java:
```diff
@@ -14,6 +17,8 @@
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
 public class CategoryReadIntegrationTest extends CategoryIntegrationTest {
+    @Autowired
+    private EntityManager entityManager;
```
사용하지 않는 것 같습니다.

gomster96 replied:
맞습니다. 감사합니다.

KJunseo reviewed (changes requested):
의견 나누고 싶은 부분이 보여서 코멘트 남깁니다!
확인 후 리뷰요청 부탁드려요!

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -55,6 +61,23 @@ private void checkCategoryExist(String name) {
+            List<ShopCategory> shopCategories = category.getShopCategories();
+            this.deleteAllShopOfCategory(shopCategories);
+        }
+        category.delete();
+    }
```
최상위 카테고리가 아닌 경우는 shopCategory를 삭제 안해주나요?

gomster96 replied:
아 모두 해줄 예정입니다. 제가 잊고있었습니다! 감사합니다.

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -55,6 +61,23 @@ private void checkCategoryExist(String name) {
+        category.delete();
+    }
+
+    @Transactional
+    public void deleteAllShopOfCategory(List<ShopCategory> shopCategories) {
+        shopCategories.forEach(shopCategory -> {
```
호출하는 메서드에 @Transactional 어노테이션이 있으면, @Transactional이 proxy기반이여서, 내부 메서드의 @Transactional 어노테이션이 호출이 안되는 것으로 알고 있습니다.

http://ojc.asia/bbs/board.php?bo_table=LecSpring&wr_id=500

gomster96 replied:
오... 감사합니다! 함부러 내부 메서드를 호출하면 안되겠군요 ..! 👍

PyeongGangKim reviewed (changes requested):
고생 많으셨습니다!! 리뷰 남겼습니다:)
궁금한 점도 남겼고, 같이 고민할 부분도 남겼습니다 확인 부탁드려요~

gomster96 committed:
refactor[#7] : 삭제 시 메세지를 보내 각 객체 내부에서 삭제 관련 로직이 일어나도록 구현

gomster96 committed:
feat[#7] : 카테고리가 삭제될 때 shopcategory 도 삭제되도록 구현

- 이때 root category의 경우 shop까지 같이 삭제되도록 구현

gomster96 committed:
feat[#7] : Shop에서 ShopCategory 관계 설정시 반대도 관계 설정

- shop에서 ShopCategory 관계 설정할 때 category와 ShopCategory의 관계 설정이 없었으므로 이를 해주었다.

gomster96 committed:
test[#7] : category delete 테스트 추가 

- category가 삭제 될 때 shop category 도 모두 삭제되는지 확인하는 테스트를 추가했다.

gomster96 committed:
fix[#7] : transaction 내부 메소드 호출 관련 

- 내부메소드를 호출하면 해당 내부메소드의 transaction 이 있으면 transaction 이 안되므로 다시 바꿔줌

gomster96 committed:
refactor[#7] : 포매팅 적용

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -34,12 +34,14 @@ public Long create(String categoryName) {
+        if (Objects.isNull(categoryCreateRequestDto.getParentId())) {
+            return this.create(categoryCreateRequestDto.getName());
+        }
+        Category parent = categoryRepository.findById(categoryCreateRequestDto.getParentId())
```
CateogryService에 아래와 같은 코드가 있는 것 같아요.
```java
    @Transactional(readOnly = true)
    public Category findById(Long categoryId) {
        return categoryRepository.findById(categoryId)
                                 .orElseThrow(CategoryNotFoundException::new);
    }
```
이 메서드를 사용하지 않고 동일한 로직을 따로 구현하신 이유가 있나요?

gomster96 replied:
네네 저번에 평강님이 이야기해주었는데, @Tansactional어노테이션이 붙어있는 내부메서드를 사용하면 Transaction이 적용되지 않는다는 것을 알게되어 Transaction을 적용하기 위해 변경하였습니다. https://mommoo.tistory.com/92

KJunseo replied:
하지만 findById를 호출하는 delete 메서드 자체에 @Transactional이 붙어있어서 옵션에 따라 트랜잭션처리가 적용됩니다.

KJunseo replied:
위의 링크에서 설명하는 부분은 @Transactional이 붙어있지 않은 메서드에서 @Transactional이 붙어있는 메서드를 호출할 때의 문제입니다. @Transactional이 붙어있는 메서드에서 @Transactional이 붙어 있는 메서드를 호출하면 propagation 옵션에 따라 부모 트랜잭션을 이어가거나 별도의 트랜잭션으로 처리하게 됩니다.

KJunseo replied:
@Transactional은 Proxy 기반이며 AOP로 구성되어 있습니다. 즉 method가 실행되기 전 트랜잭션을 묶게 되는데, 스프링에서는 인스턴스에서 처음 호출하는 메서드나 클래스의 속성을 따라가게 됩니다. 따라서 처음 호출하는 메서드에 @Transactional이 있다면 하위 메서드에도 적용됩니다.

PyeongGangKim replied:
앗,,,제가 쓴 글은 그 의미가 아니였는데,, 제가 의미 전달을 잘 못했군요,,죄송합니다

gomster96 replied:
아 그렇군요...! 제가 나름 찾아봤는데 이해가 부족했네요!! 그리고 빠르게 답변해주셔서 정말 감사합니다.!! 저도 바로 봤어야하는뎁... ㅜㅜ 저도 사실 바꾸면서 이게 맞나? 싶은 생각이 들었는데, 더 꼼꼼히 봤어야했네요!!  감사합니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/category/application/CategoryService.java:
```diff
@@ -55,6 +57,13 @@ private void checkCategoryExist(String name) {
 
+    @Transactional
+    public void delete(CategoryRequestDto categoryRequestDto) {
+        Category category = categoryRepository.findById(categoryRequestDto.getId())
+                                              .orElseThrow(CategoryNotFoundException::new);
```
기존에 존재하는 findById()를 사용하지 않으시는 이유가 있나요?

gomster96 replied:
위의 내용과 같습니다.

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Category.java:
```diff
@@ -66,6 +71,27 @@ public boolean isSameName(String name) {
+            throw new CategoryAlreadyDeletedException();
+        }
+
+        deleteShopCategory(Objects.isNull(this.parent));
```
로직을 읽기가 조금 힘들었어요. shopCategory를 지우라는 건 알겠는데 갑자기 Null 체크가 나와서 이해하는데 조금 걸렸습니다!
isParent()나 isRoot()와 같이 메서드로 표현하는 건 어떻게 생각하시나요?

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/category/ShopCategory.java:
```diff
@@ -27,4 +28,15 @@ public ShopCategory(Shop shop, Category category) {
+    public void delete(boolean isRootCategory) {
+        this.deleteContent();
+        if (isRootCategory) {
+            this.shop.delete();
```
shop.delete()를 보니 내부에서 shop.deleteContent()만 호출하고 있던데 나눠서 사용하는 이유가 있을까요?

gomster96 replied:
나름대로 확장가능성을 고려하여 shop의 일은 shop안에서 처리하도록 구현 했던 것이였는데, 로직 이해만 복잡하게 하는 부분일까요?

KJunseo commented on backend/src/main/java/com/handong/rebon/category/domain/Category.java:
```diff
@@ -66,6 +71,27 @@ public boolean isSameName(String name) {
+
+        deleteShopCategory(Objects.isNull(this.parent));
+
+        this.children.delete();
```
하위 카테고리 자신 뿐만아니라 하위 카테고리와 연관된 shopCategory도 삭제 처리해야하지 않나요??

gomster96 replied:
네 맞습니다...! 흠 구현하겠습니다!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/Shop.java:
```diff
@@ -92,10 +92,11 @@ public void addTags(List<Tag> tags) {
 
     public void addCategories(Category parent, List<Category> subCategories) {
         this.category = parent;
+        parent.addShopCategory(new ShopCategory(this, parent));
```
이 부분에 대해 생각을 해봤는데 저도 객체내에서 처리할 수 있다면 그게 좋을 것 같기는 해요. 하지만 그때 말한 것 처럼 현재 구조에서는 Shop이 상위 카테고리 필드를 가지고 있어서 shop의 필드로 한 번, shopCategory로 한 번 두군데의 관리 포인트가 생기게 되는 것 같아요. 따라서 이 문제를 없애기 위해서는

1) 기존 처럼 상위 카테고리는 shopCategory를 만들지 않고, repository를 통해 삭제
2) 상위 카테고리도 shopCategory로 만들기. 이때 shop의 category 필드는 삭제.(이 필드를 사용하는 로직은 전부 수정해야할듯)

이 두 가지 방법이 있을 것 같은데 이 부분에 대해 의견 나눠보고 싶네요.

PyeongGangKim replied:
개인적으로는 2번 방법이 좀 더 괜찮을 거 같다는 생각이 들어요!! 삭제할 때도 로직들이 좀 더 자연스러울 거 같아요!!

gomster96 replied:
음 저도 일단은 2번이 좋은것 같네요..! 고칠게 좀  많더라도 이렇게 하는게 평강님 말대로 로직들이 자연스러울 것 같아요.

KJunseo reviewed (changes requested):
저번에 이야기 나눴던 부분에 대해 제 생각을 적어봤어요. 테스트 부분은 결론에 따라 달라질 수 있을 것 같아서 프로덕션 단의 코드 리뷰만 진행했습니다! 확인해보시고 다시 요청주세요.

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/category/domain/Categories.java:
```diff
@@ -23,4 +25,8 @@ public void addChild(Category category) {
     private void checkDuplicateCategory(Category category) {
         this.categories.forEach(child -> child.validateSame(category));
     }
+
+    public void delete(){
+        this.categories.forEach(BaseEntity::deleteContent);
+    }
```
이 부분에서
`this.categories.forEach(Category::delete)`
를 사용하면, 하위 카테고리들이랑 카테고리들과 연관된 shopCategory도 삭제를 할 수 있지 않을까요??

gomster96 replied:
맞습니다..! 이렇게 하려고 했던 것이였는데 ㅋㅋㅋㅋ 구현하다가 놓쳤네요 감사합니다.

PyeongGangKim reviewed (changes requested):
고생많으셨습니다:)
카테고리가 계층 구조로 되어 있어, 삭제하는 부분이 쉽지 않은 거 같네요!!
리뷰 남긴 거 확인 부탁드려요~!

gomster96 committed:
refactor[#7] : 리뷰내용 반영

gomster96 committed:
fix[#7] : 루트카테고리가 삭제되면 해당 카테고리를 가지고 있는 모든 shop 삭제

- 효율적인 쿼리를 위해 categoryService에서 deleteAllInBatch 를 쓰고 싶었지만 참조무결성 오류가 자꾸 발생하여 사용하지 못하였다.... (찾는중)

gomster96 committed:
refactor[#7] : test 관련 수정

gomster96 committed:
feat[#7] : delete 할 때 다른 연관관계 있는 컬럼들도 soft delete 되도록 처리

PyeongGangKim commented on backend/src/main/java/com/handong/rebon/shop/domain/repository/ShopRepository.java:
```diff
@@ -1,8 +1,14 @@
 public interface ShopRepository extends JpaRepository<Shop, Long>, ShopRepositoryCustom {
+
+    @Query("select s from Shop s where s.category.id = :id")
+    List<Shop> findShopsByCategoryId(Long id);
```
이 부분 카테고리를 넘겨주면, Jpql을 따로 만들지 않아도 되지 않을까요?

gomster96 replied:
맞습니다. 그렇게 하는게 훨씬 좋을것 같네요 감사합니다!

PyeongGangKim reviewed (approved):
리뷰는 하나 남겼어요!! 어푸르브 하겠습니다~!!

KJunseo commented on backend/src/main/java/com/handong/rebon/shop/domain/repository/ShopRepository.java:
```diff
@@ -1,8 +1,14 @@
 public interface ShopRepository extends JpaRepository<Shop, Long>, ShopRepositoryCustom {
+
+    @Query("select s from Shop s where s.category.id = :id")
+    List<Shop> findShopsByCategoryId(Long id);
```
실행이 안되어서 찾아보니 @Param을 넣어줘야 실행이 제대로 되는것 같습니다!

KJunseo reviewed (approved):
고생하셨습니다! 리뷰를 꽤 많이 드린 것 같네요.
마지막으로 실제로 실행이 안되는 것 같아서 해당 부분에 관한 코멘트만 남겼습니다!
어푸르트 할테니 이부분만 확인해주세요

gomster96 committed:
refactor[#7] : 리뷰반영

jpql 쿼리문을 굳이 써주지 않고 쉽게 쓸 수 있도록 구현

gomster96 committed:
Merge branch 'dev' into be/delete-category/#7
